frontend/src/modules/addonpromotion/components/AddonPromotion.tsx

```diff
@@ -39,13 +46,33 @@ export class AddonPromotionBase extends React.Component<InternalProps, State> {
         window.removeEventListener('message', this.handleMessages);
     }
 
+    // Hide Add-On Promotion if Add-On installed while active
     handleMessages: (event: MessageEvent) => void = (event: MessageEvent) => {
-        const data = JSON.parse((event.data as any) as string);
-        if (data._type === 'PontoonAddonInfo') {
-            if (data.value.installed) {
-                this.setState({
-                    installed: true,
-                });
+        // only allow messages from authorized senders (extension content script, or Pontoon itself)
+        if (event.origin !== window.origin || event.source !== window) {
+            return;
+        }
+        let data: PontoonAddonInfoMessage | undefined;
+        switch (typeof event.data) {
+            case 'object':
+                data = event.data as PontoonAddonInfoMessage;
+                break;
+            case 'string':
+                // backward compatibility
+                // TODO: remove some reasonable time after https://github.com/MikkCZ/pontoon-addon/pull/155 is released
+                // and convert this switch into a condition
+                try {
+                    data = JSON.parse(
+                        event.data as string,
+                    ) as PontoonAddonInfoMessage;
+                } catch (_) {
+                    return;
+                }
+                break;
+        }
+        if (data && data._type === 'PontoonAddonInfo' && data.value) {
```

Comment:
Can you use optional chaining here and below?

```
if (data?._type === 'PontoonAddonInfo' && data.value?.installed === true)
```

I doubt we can do so in our js code base yet, but in the TS one, this should be transpiled away.

Reply:
It's done. I touched another place in the code as well, that deserved to be polished after migration to TypeScript.

Reply:
Sweet, thanks for removing the ts-expect-error.